backend/metasfresh-webui-api/src/main/java/de/metas/ui/web/pickingV2/productsToPick/process/ProductsToPickHelper.java

```diff
@@ -0,0 +1,134 @@
+import de.metas.handlingunits.picking.PickingCandidateService;
+import de.metas.handlingunits.picking.candidate.commands.PickHUResult;
+import de.metas.handlingunits.picking.requests.PickRequest;
+import de.metas.ui.web.pickingV2.config.PickingConfigRepositoryV2;
+import de.metas.ui.web.pickingV2.config.PickingConfigV2;
+import de.metas.ui.web.pickingV2.productsToPick.rows.ProductsToPickRow;
+import de.metas.ui.web.pickingV2.productsToPick.rows.ProductsToPickRowsService;
+import de.metas.ui.web.window.datatypes.DocumentId;
+import de.metas.util.GuavaCollectors;
+import lombok.NonNull;
+import org.adempiere.util.lang.ImmutablePair;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Stream;
+
+@Component
+class ProductsToPickHelper
+{
+	private final PickingCandidateService pickingCandidateService;
+	private final ProductsToPickRowsService productsToPickRowsService;
+
+	private final PickingConfigRepositoryV2 pickingConfigRepo;
+	private PickingConfigV2 _pickingConfig; // lazy
+
+	protected final PickingConfigV2 getPickingConfig()
+	{
+		PickingConfigV2 pickingConfig = _pickingConfig;
+		if (pickingConfig == null)
+		{
+			pickingConfig = _pickingConfig = pickingConfigRepo.getPickingConfig();
+		}
+		return pickingConfig;
+	}
+
+	ProductsToPickHelper(final PickingCandidateService pickingCandidateService,
+			final ProductsToPickRowsService productsToPickRowsService,
+			final PickingConfigRepositoryV2 pickingConfigRepo)
+	{
+		this.pickingCandidateService = pickingCandidateService;
+		this.productsToPickRowsService = productsToPickRowsService;
+		this.pickingConfigRepo = pickingConfigRepo;
+	}
+
+	@NonNull
+	public ImmutableList<ImmutablePair<DocumentId, PickingCandidate>> pick(final List<ProductsToPickRow> selectedRows)
+	{
+		return streamRowsEligibleForPicking(selectedRows)
+				.map(row -> {
+					final PickHUResult result = pickingCandidateService.pickHU(createPickRequest(row));
+					return ImmutablePair.of(row.getId(), result.getPickingCandidate());
+				})
+				.collect(GuavaCollectors.toImmutableList());
+	}
+
+	public ImmutableList<ImmutablePair<DocumentId, PickingCandidate>> setPackingInstruction(final List<ProductsToPickRow> selectedRows, final HuPackingInstructionsId huPackingInstructionsId)
+	{
+
+		final Map<PickingCandidateId, DocumentId> rowIdsByPickingCandidateId = streamRowsEligibleForPacking(selectedRows)
+				.collect(ImmutableMap.toImmutableMap(ProductsToPickRow::getPickingCandidateId, ProductsToPickRow::getId));
+
+		final Set<PickingCandidateId> pickingCandidateIds = rowIdsByPickingCandidateId.keySet();
+		final List<PickingCandidate> pickingCandidates = pickingCandidateService.setHuPackingInstructionId(pickingCandidateIds, huPackingInstructionsId);
+
+		return pickingCandidates.stream()
+				.map(cand -> ImmutablePair.of(rowIdsByPickingCandidateId.get(cand.getId()), cand))
+				.collect(GuavaCollectors.toImmutableList());
```

Comment:
minor: u can use ImmutableList.toImmutableList()

Reply:
Pretty sure i looked for that one and missed it. no idea how.

Anyway, done.